nft/ethereum_payment_obligation.go

```diff
@@ -143,79 +129,113 @@ func (s *ethereumPaymentObligation) MintNFT(ctx context.Context, documentID []by
-		TokenID:       requestData.TokenID.String(),
+		TokenID:       tokenID.String(),
 	}, nil
 }
 
-// OwnerOf returns the owner of the NFT token on ethereum chain
-func (s *ethereumPaymentObligation) OwnerOf(registry common.Address, tokenID []byte) (owner common.Address, err error) {
-	contract, err := s.bindContract(registry, s.ethClient)
-	if err != nil {
-		return owner, errors.New("failed to bind the registry contract: %v", err)
-	}
+func (s *ethereumPaymentObligation) minter(ctx context.Context, tokenID TokenID, documentID []byte, registryAddress, depositAddress string, proofFields []string) func(accountID identity.CentID, txID uuid.UUID, txMan transactions.Manager, errOut chan<- error) {
+	return func(accountID identity.CentID, txID uuid.UUID, txMan transactions.Manager, errOut chan<- error) {
+		tc, err := contextutil.Account(ctx)
+		if err != nil {
+			errOut <- err
+		}
 
-	opts, cancF := s.ethClient.GetGethCallOpts(false)
-	defer cancF()
+		// registry address
+		registry := common.HexToAddress(registryAddress)
 
-	return contract.OwnerOf(opts, utils.ByteSliceToBigInt(tokenID))
-}
+		model, err := s.docSrv.GetCurrentVersion(ctx, documentID)
+		if err != nil {
+			errOut <- err
+		}
 
-// sendMintTransaction sends the actual transaction to mint the NFT
-func (s *ethereumPaymentObligation) sendMintTransaction(
-	cid identity.CentID,
-	contract ethereumPaymentObligationContract,
-	opts *bind.TransactOpts,
-	requestData MintRequest,
-	registry common.Address,
-	documentID []byte) (uuid.UUID, error) {
+		cd, err := model.PackCoreDocument()
+		if err != nil {
+			errOut <- err
+		}
 
-	r := requestData.copy()
+		data := cd.EmbeddedData
+		cd, err = coredocument.PrepareNewVersion(*cd, nil)
+		if err != nil {
+			errOut <- err
+		}
 
-	// Run within transaction
-	// We use context.Background() for now so that the transaction is only limited by ethereum timeouts
-	txID, _, err := s.txManager.ExecuteWithinTX(context.Background(), cid, uuid.Nil, "Minting NFT", func(accountID identity.CentID, txID uuid.UUID, txMan transactions.Manager, errOut chan<- error) {
-		ethTX, err := s.ethClient.SubmitTransactionWithRetries(contract.Mint, opts, r.To, r.TokenID, r.TokenURI, r.AnchorID,
-			r.MerkleRoot, r.Values, r.Salts, r.Proofs)
+		cd.EmbeddedData = data
+		err = coredocument.AddNFTToReadRules(cd, registry, tokenID[:])
 		if err != nil {
 			errOut <- err
 		}
 
-		log.Infof("Sent off ethTX to mint [tokenID: %x, anchor: %x, registry: %x] to payment obligation contract. Ethereum transaction hash [%x] and Nonce [%v] and Check [%v]",
-			requestData.TokenID, requestData.AnchorID, requestData.To, ethTX.Hash(), ethTX.Nonce(), ethTX.CheckNonce())
-		log.Infof("Transfer pending: 0x%x\n", ethTX.Hash())
+		model, err = s.docSrv.DeriveFromCoreDocument(cd)
+		if err != nil {
+			errOut <- err
+		}
 
-		res, err := ethereum.QueueEthTXStatusTask(accountID, txID, ethTX.Hash(), true, s.queue)
+		_, _, err = s.docSrv.Update(contextutil.WithTX(ctx, txID), model)
```

Comment:
Doing this rule update and anchoring before minting is the main change

Reply:
As discussed in person. We need to wait for the document update because the latest current version is needed and the tokenID needs to be part of the merkle proof.